Add `prefer-modern-dom-apis` rule

mhatvan opened the pull request:
Fixes #344.

The rule can be tested in this gist here: https://astexplorer.net/#/gist/338d492213d95b40122046fb29abd1bc/bb5030be4689f651524402efad8d24d2b310d588

The links in the description for `prepend()` and `append()` from MDN are returning 404 by the way, they should be updated.


<!-- Issuehunt content -->

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#344: Rule proposal:  `prefer-modern-dom-apis`](https://issuehunt.io/repos/55832243/issues/344)
---

IssueHunt has been backed by the following sponsors. [Become a sponsor](https://issuehunt.io/membership/members)
</details>
<!-- /Issuehunt content-->

MrHen commented on rules/prefer-modern-dom-apis.js:
```diff
@@ -0,0 +1,121 @@
+		}
+
+		return null;
+	};
```
Move this helper function out of `checkForReplaceChildOrInsertBefore`. There may be a suitable util here: https://github.com/sindresorhus/eslint-plugin-unicorn/tree/master/rules/utils. Perhaps `is-method-named`?

Also, `args` is not a very good argument name. Perhaps `nodeArgument`?

mhatvan replied:
`is-method-named` would not help me, since i cannot say what the argument name will be.
There are multiple other rules with `const args = node.arguments;` and their own scoped helper functions, e.g: `prefer-node-remove` in the code already, but I am also fine with renaming it ;)

MrHen commented on rules/prefer-modern-dom-apis.js:
```diff
@@ -0,0 +1,121 @@
+	const identifierName = node.callee.property.name;
+
+	// Only handle methods that exist in forbiddenIdentifierNames
+	if (forbiddenIdentifierNames.has(identifierName)) {
```
I prefer the return-early when the condition is being used as a guard.

```
	if (!forbiddenIdentifierNames.has(identifierName)) {
		return;
	}
```

mhatvan replied:
Good point, changing it in the upcoming commit.

MrHen commented on rules/prefer-modern-dom-apis.js:
```diff
@@ -0,0 +1,121 @@
+	const forbiddenIdentifierNames = new Map([
+		['replaceChild', 'replaceWith'],
+		['insertBefore', 'before']
+	]);
```
Move this to a constant variable outside of `checkForReplaceChildOrInsertBefore`.

mhatvan replied:
Do you mean the global scope?
Why would I move it outside when it is only used in `checkForReplaceChildOrInsertBefore`?
The more scoped the better imo

MrHen replied:
We recently introduced a rule that will enforce this behavior. See the discussion here: https://github.com/sindresorhus/eslint-plugin-unicorn/issues/230.

The short version is that there isn't any significant benefit to scoping the function and a couple minor benefits to code readability, maintainability and performance if you move it out.

MrHen commented on rules/prefer-modern-dom-apis.js:
```diff
@@ -0,0 +1,121 @@
+		}
+
+		return null;
+	};
```
Same comment as before. This should be moved to the parent scope.

MrHen commented on rules/prefer-modern-dom-apis.js:
```diff
@@ -0,0 +1,121 @@
+		['afterbegin', 'prepend'],
+		['beforeend', 'append'],
+		['afterend', 'after']
+	]);
```
Same comment as before. This should be moved to the parent scope as a constant.

MrHen commented on rules/prefer-modern-dom-apis.js:
```diff
@@ -0,0 +1,121 @@
+	if (
+		identifierName === 'insertAdjacentText' ||
+		identifierName === 'insertAdjacentElement'
+	) {
```
Same comment as before. Use return-early pattern.

MrHen commented on rules/prefer-modern-dom-apis.js:
```diff
@@ -0,0 +1,121 @@
+		CallExpression(node) {
+			if (
+				node.callee.type === 'MemberExpression' &&
+				node.arguments.length === 2
```
You can include these checks in the selector pattern for the rule. Not required but in case you hadn't seen it yet, here are the docs: https://eslint.org/docs/developer-guide/selectors

MrHen commented on test/prefer-modern-dom-apis.js:
```diff
@@ -0,0 +1,129 @@
+			output: 'referenceNode.after(newNode);'
+		}
+	]
+});
```
Add tests for the following scenarios:

```
referenceNode.insertAdjacentElement(foo, bar);

referenceNode.insertAdjacentElement(
    "afterend",
    newNode
);

referenceNode.insertAdjacentElement( // inline comments
    "afterend", // inline comments
    newNode,  // inline comments
); // inline comments

parentNode.replaceChild(
    newChildNode,
    oldChildNode
);

parentNode.replaceChild( // inline comments
    newChildNode, // inline comments
    oldChildNode // inline comments
);

parentNode.insertBefore(alfa, beta).insertBefore(charlie, delta);
```

mhatvan replied:
Cool, thanks for these test scenario suggestions!
So I guess the inline comments should be preserved? How can I do that?

MrHen replied:
It's a total pain. If you want, you can disable the fixer if there are comments in the node. Take a look at `getCommentsInside` or `commentsExistBetween`.

https://eslint.org/docs/developer-guide/working-with-rules#contextgetsourcecode

mhatvan replied:
`referenceNode.insertAdjacentElement(foo, bar);` -> this is incorrect syntax, I would not touch it and not modify it and add it to the valid array, correct?

`parentNode.insertBefore(alfa, beta).insertBefore(charlie, delta);` -> how to handle this?

MrHen replied:
`referenceNode.insertAdjacentElement(foo, bar);` should be just as valid as `referenceNode.insertAdjacentElement("beforebegin", newNode);`. It's just using a variable instead of a literal.

----

`parentNode.insertBefore(alfa, beta).insertBefore(charlie, delta);` cannot be fixed properly because the return value is being used. A similar test would be `var foo = parentNode.insertBefore(alfa, beta);`

mhatvan commented on readme.md:
```diff
@@ -142,19 +140,17 @@ See the [ESLint docs](http://eslint.org/docs/user-guide/configuring#extending-co
 - [Adam Babcock](https://github.com/MrHen)
 
-###### Former
+### Former
```
In this case, these were intended, manual changes because markdownlint threw me a couple of warnings e.g. https://github.com/DavidAnson/markdownlint/blob/v0.16.0/doc/Rules.md#md001

But I am fine with reverting back to how it was before, however ;)

MrHen replied:
I'm okay making the changes in a separate PR.

mhatvan commented:
@MrHen can you take another look at the changes you requested? I implemented your recommendations a while ago.

MrHen commented:
@mhatvan Sorry, I'm actually in the middle of changing employers right now. I probably won't get a chance to look at it in the next couple weeks.

fisker commented on readme.md:
```diff
@@ -8,14 +8,12 @@ You might want to check out [XO](https://github.com/xojs/xo), which includes thi
 
 [**Propose or contribute a new rule ➡**](.github/contributing.md)
 
-
```
restore blank lines in this file please

mhatvan replied:
More than one blank line is bad practice, especially since it doesn't affect the layout at all.

fisker commented on readme.md:
````diff
@@ -8,14 +8,12 @@ You might want to check out [XO](https://github.com/xojs/xo), which includes thi
 ## Install
 
-```
+```bash
````
restore this,since there is `$` inside

fisker replied:
there is `$` inside

mhatvan replied:
wouldn't it be better to have it conforming to markdownlint? 

https://github.com/DavidAnson/markdownlint/blob/master/doc/Rules.md#md040

sindresorhus replied:
```suggestion
```console
```

works better

fisker commented on index.js:
```diff
@@ -13,9 +13,7 @@ module.exports = {
-			plugins: [
-				'unicorn'
-			],
+			plugins: ['unicorn'],
```
```suggestion
			plugins: [
				'unicorn'	
			],
```

fisker commented on rules/prefer-modern-dom-apis.js:
```diff
@@ -0,0 +1,136 @@
+'use strict';
+const getDocsUrl = require('./utils/get-docs-url');
```
rebase this PR, 
run `yarn lint rules/prefer-modern-dom-apis.js --fix`
you'll get an error here, and will be fixed

fisker replied:
sorry this will not fix currectly, after rebase, rename it to `getDocumentationUrl`

https://github.com/sindresorhus/eslint-plugin-unicorn/blob/d638e5477d7240ea777dcaf249d0121a5a1910eb/rules/catch-error-name.js#L4

fisker commented:
since [`childNode.replaceWith()`](https://developer.mozilla.org/en-US/docs/Web/API/ChildNode/replaceWith) and [`Node.replaceChild()`](https://developer.mozilla.org/en-US/docs/Web/API/Node/replaceChild) returns different value

also `ChildNode.before()`

[`ChildNode.before()`](https://developer.mozilla.org/en-US/docs/Web/API/ChildNode/before)
[`Node.insertBefore()`](https://developer.mozilla.org/en-US/docs/Web/API/Node/insertBefore)
[`Element.insertAdjacentElement()`](https://developer.mozilla.org/en-US/docs/Web/API/Element/insertAdjacentElement)

so, maybe you want do a similar check like https://github.com/sindresorhus/eslint-plugin-unicorn/pull/321

mhatvan commented:
@fisker I pushed some code adaptions in the newest commit.

fisker commented:
what about this?

> since [`childNode.replaceWith()`](https://developer.mozilla.org/en-US/docs/Web/API/ChildNode/replaceWith) and [`Node.replaceChild()`](https://developer.mozilla.org/en-US/docs/Web/API/Node/replaceChild) returns different value
> 
> also `ChildNode.before()`
> 
> [`ChildNode.before()`](https://developer.mozilla.org/en-US/docs/Web/API/ChildNode/before)
> [`Node.insertBefore()`](https://developer.mozilla.org/en-US/docs/Web/API/Node/insertBefore)
> [`Element.insertAdjacentElement()`](https://developer.mozilla.org/en-US/docs/Web/API/Element/insertAdjacentElement)
> 
> so, maybe you want do a similar check like #321

mhatvan commented:
> so, maybe you want do a similar check like #321
you linked the complete PR without specifying what you exactly want me to change, I need more details here.

fisker commented:
> you linked the complete PR without specifying what you exactly want me to change, I need more details here.

@mhatvan I mean some methods returns different result, so we need check if the return value is used.

sindresorhus commented on docs/rules/prefer-modern-dom-apis.md:
```diff
@@ -0,0 +1,52 @@
+- `childNode.replaceWith(newNode);` over `parentNode.replaceChild(newNode, oldNode)`;
+- `referenceNode.before(newNode);` over `parentNode.insertBefore(newNode, referenceNode);`
+- `referenceNode.before("text");` over `referenceNode.insertAdjacentText("beforebegin", "text");`
+- `referenceNode.before(newNode);` over `referenceNode.insertAdjacentElement("beforebegin", newNode);`
```
Can you linkify all these to their MDN page?

mhatvan commented:
@sindresorhus I implemented the changes you requested, can we get this merged now?

I still don't understand @fisker feedback:
> childNode.replaceWith() and Node.replaceChild() returns different value

I implemented it according to the examples in MDN and they return the expected result.

sindresorhus commented:
@mhatvan `Node.replaceChild()` returns a value. If the user actually uses this value, we cannot safely auto-fix with with `ChildNode.replaceWith()` as it doesn't return any value. Can you handle this and add a test for that?

mhatvan commented:
@sindresorhus Sorry, I missed that you already left a comment a while ago. I pushed another commit which implements the requested changes and added additional tests.

fisker commented:
@mhatvan I think we should still  report an error, just can't fix

sindresorhus commented:
>@mhatvan I think we should still  report an error, just can't fix

👍

mhatvan commented:
@sindresorhus @fisker feedback implemented, I pushed another commit ;)

sindresorhus commented:
Looks good. Thanks for contributing 👌

sindresorhus commented:
Looks good. Thanks for contributing 👌

fregante commented:
Awesome!

fisker commented on rules/prefer-modern-dom-apis.js:
```diff
@@ -0,0 +1,158 @@
+	}
+
+	return false;
+};
```
I don't think this is enough, 

simple one:

```js
console.log(referenceNode.insertAdjacentElement("beforebegin", newNode));
``
